Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift

```diff
@@ -0,0 +1,169 @@
+  func audioVideoPlayer(_ player: AudioVideoPlayer, didProgressToTime seconds: Double) {}
+}
+
+public class AudioVideoPlayer: AVPlayer {
+  public static let defaultTimescale = CMTimeScale(10)
+
+  public private(set) lazy var playbackInterval: (startAt: Double, endAt: Double) = (0, duration)
+  public var allowLooping = true
+  public var periodicTimeObserverTimeInterval: CMTime = .init(value: 1, timescale: AudioVideoPlayer.defaultTimescale) {
+    didSet {
+      removeTimeObserver()
+      setupTimeObserver()
+    }
+  }
+  public weak var delegate: AudioVideoPlayerDelegate?
+  
+  private var timeObserver: Any?
+  
+  override public init() {
+    super.init()
+    setupTimeObserver()
+  }
+
+  public init(
+    url: URL,
+    periodicTimeObserverTimeInterval timeInterval: CMTime = CMTimeMake(value: 1, timescale: AudioVideoPlayer.defaultTimescale)
+  ) {
+    super.init(url: url)
+    setupTimeObserver()
+  }
+
+  public init(
+    playerItem item: AVPlayerItem?,
+    periodicTimeObserverTimeInterval timeInterval: CMTime = CMTimeMake(value: 1, timescale: AudioVideoPlayer.defaultTimescale)
+  ) {
+    super.init(playerItem: item)
+    setupTimeObserver()
+  }
+  
+  convenience public init?(urlString: String) {
+    guard let url = URL(string: urlString) else { return nil }
+    self.init(url: url)
+  }
+}
+
+// MARK: - API
+public extension AudioVideoPlayer {
+  override func play() {
+    super.play()
+    delegate?.audioVideoPlayer(didBeginPlayback: self)
+  }
+
+  override func pause() {
+    super.pause()
+    delegate?.audioVideoPlayer(didPausePlayback: self)
+  }
+
+  override func replaceCurrentItem(with item: AVPlayerItem?) {
+    super.replaceCurrentItem(with: item)
+    playbackInterval = (0, duration)
+    removeTimeObserver()
+    setupTimeObserver()
+  }
+
+  /// Set `playbackInterval` and force player to start playing from the `startAt` timestamp.
+  func setPlaybackInterval(
+    startAt: Double,
+    endAt: Double,
+    timescale: CMTimeScale = AudioVideoPlayer.defaultTimescale
+  ) {
+    guard startAt < endAt else { fatalError("`startAt` should be less than `endAt`") }
+    setPlaybackPosition(startAt: startAt)
+    playbackInterval = (startAt, endAt)
+  }
+
+  func setPlaybackPosition(
+    startAt: Double,
+    timescale: CMTimeScale = AudioVideoPlayer.defaultTimescale
+  ) {
+    removeTimeObserver()
+    seek(to: CMTime(seconds: startAt, preferredTimescale: timescale)) { _ in
+      self.setupTimeObserver()
+    }
+  }
+
+  /// Update `playbackInterval` and keep playing from the current location.
+  func updatePlaybackInterval(startAt: Double, endAt: Double) {
+    guard startAt < endAt else { fatalError("`startAt` should be less than `endAt`") }
+    playbackInterval = (startAt, endAt)
+    removeTimeObserver()
+    setupTimeObserver()
+  }
+
+  var duration: Double {
+    currentItem?.asset.duration.seconds ?? 0
+  }
+  
+  var isPlaying: Bool {
+    rate != 0.0
+  }
```

Comment:
It seems that it's possible for the player to stall due to buffer issues, etc. In that case the rate seems to not go to 0. We should also check if there's any error.

```suggestion
  var isPlaying: Bool {
    rate != 0.0 && error == nil
  }
```

Reply:
Used `status` instead, but very good point.